src/hyp_map.rs

```diff
@@ -476,3 +504,103 @@ impl UmodeSlotMapper<'_> {
             .map_err(|_| Error::MapFailed)
     }
 }
+
+/// Umode Buffer errors.
+#[derive(Debug)]
+pub enum UmodeBufferError {
+    /// Address overflow on increment.
+    AddressOverflow,
+    /// Buffer is full.
+    NoSpace,
+}
+
+/// The U-mode buffer is a private area of the page table used by the hypervisor to pass data to the
```

Comment:
For my understanding: Can you explain what data items we're expecting to pass via this mechanism? Do we have additional static knowledge, e.g. that the data must be of one of a finite number of types?

(background: having seen only this patch, I'm unconvinced that we need a fully general buffer mechanism)

Reply:
The data type will be call specific. I've been going back and forth between copying a repr(C) structure or having a byte buffer, and I went for the second for two reasons:

1. In the immediate need, the easiest way to get the data needed for attestation (PCRs and CDI_ID) is that of the byte buffer.
2. Nothing will prevent to expand the interface and just be able to append a particular type.

This is not a generic buffer, at least on this side. It's the simplest mechanism that will allow -- with fewer moving parts -- to implement a passing of data from salus to u-mode that I could think of.

Reply:
(Edited and then deleted the wrong comment by mistake. Rewriting it)
This will be call specific. I have been going back and forth whether to have a repr(c) structure in a shared memory or going with byte buffer, I went for the second for two main reasons:

1. Given the immediate need of this, the best way to get data for attestation, PCRs and CDI ID, is through byte buffers.
2. In case things change, it will be easy to just expand this to be able to append (copy) a specific type.

This is not a generic buffer, at least not on the hypervisor side. This was the simplest way (fewer moving parts in Salus) to get data to U-mode.

Even the append() function, which might appear odds, serves the purpose of being able to collect all data and write them in order to the buffer.

Reply:
I'm probably making incorrect assumptions on how this is to be used and as a result I'm asking weird questions... Let me clarify where I'm coming from:

At a very high level, I was assuming that all you need is to encode an operation type and the data associated with it, dump that somewhere in memory. The umode can decode and process it. The (IMHO) natural way of doing this in rust is defining an enum type with each case holding relevant data. Then you can use a suitable serialization mechanism (internal interface with rust on either end, so something like bincode would probably work well) to encode the request. That's my mental model at least, probably overly simplistic.

What you have in the test code gave me the impression that the request handlers will have to do their own interpretation of what they find in the buffer (that's what I meant by "general buffer mechanism"), but I really don't know how you imagine actual operations to look like, so perhaps you can just sketch how an attestation-related operation would work?

Reply:
Fair enough, that approach makes sense. Please don't read these comments as requests to change something, I'm really just asking questions on stuff that seems unintuitive to me :)